File: server/src/types/solves.go
package types

type Move struct {
	Move string  `json:"move"`
	Time float32 `json:"time"`
}

type Solve struct {
	Id        int     `json:"id"`
	Time      float32 `json:"time"`
	Scramble  string  `json:"scramble"`
	Moves     string  `json:"moves"`
	Puzzle    string  `json:"puzzle"`
	SessionId int     `json:"sessionId"`
}

File: ui/src/lib/scripts/api/solves.ts
import { post } from "./common";

export type MinSolve = {
    id: number;
    time: number;
    penalty: Penalty;
}

export type Penalty = undefined | "+2" | "DNF";

export type Solve = 
    MinSolve & {
    scramble: string;
    moves: string;
    puzzle: string;
    sessionId: number;
    penalty: Penalty;
}

export async function create(solve: Solve): Promise<number> {
    return post("/createSolve", solve);
}

export async function read(id: number): Promise<Solve> {
    return post("/readSolve", { id });
}

export async function readAll(sessionId: number): Promise<MinSolve[]> {
    return (await post("/readSolves", { sessionId })) ?? [];
}

export async function del(id: number): Promise<void> {
    return post("/deleteSolve", { id });
}